### PR TITLE
Sender-based synapses

### PR DESCRIPTION
**Pull request by rcoreilly**
switching back to sender-based.

### CODE REVIEW

**Comment by rcoreilly** [comment]
The last sender-based version was: https://github.com/emer/axon/commit/b60251691a43d632e7e157e62b1085744fabe08c released as version 1.7.1  on Jan 25, 2023.

File: axon/threads_test.go [hunk 1]
```diff
@@ -157,66 +148,6 @@ func TestDeterministicSingleThreadedTraining(t *testing.T) {
 	assert.False(t, netA.WtsHash() == netB.WtsHash())
 }
 
-func TestMultithreadedSendSpike(t *testing.T) {
```
Comment (siboehm) [inline axon/threads_test.go:160]: Why delete these? These are great unit tests ensuring the thread-safety of big parts of the cycle-level processing.

**Review by siboehm — APPROVED** [review]
Left some comments. Seems to make things ~30% faster on my MBP, nice! I wonder where those speedups are coming from (SynCaRecv consolidation vs SendSpike).

Just a note on the atomics, looking at the benchmark.md's you posted: Atomics scale poorly with threads (the time each atomic add takes scales linearly with threads due to cache coherency issues), so it's important to benchmark this stuff using a large number of threads.

**Review by siboehm — APPROVED** [review]
Left some comments. Seems to make things ~30% faster on my MBP, nice! I wonder where those speedups are coming from (SynCaRecv consolidation vs SendSpike).

Just a note on the atomics, looking at the benchmark.md's you posted: Atomics scale poorly with threads (the time each atomic add takes scales linearly with threads due to cache coherency issues), so it's important to benchmark this stuff using a large number of threads.

**Comment by siboehm** [comment]
Oh shit shouldn't have approved bc you had automerger on

File: axon/threads.go [hunk 2]
```diff
@@ -6,116 +6,79 @@ package axon
...
 
 	"github.com/emer/emergent/timer"
+	"github.com/goki/ki/atomctr"
 	"github.com/goki/ki/ints"
 )
 
+// Maps the given function across the [0, total) range of items, using
+// nThreads goroutines, in smaller-sized chunks for better load balancing.
+// this may be better for larger number of threads, but is not better for small N
+func ParallelChunkRun(fun func(st, ed int), total int, nThreads int) {
```
[the remark this answers is not in the document]
Comment (rcoreilly) [reply axon/threads.go:22]: It was how it was written in 1.6.16 which is significantly faster than 1.7.24 -- I didn't thoroughly test variants of this so leaving it in for future exploration.

File: axon/threads_test.go [hunk 3]
```diff
@@ -19,12 +19,11 @@ const (
...
 
-// TestMultithreadingCycleFun tests the whole net.Cycle() function, turning on
-// multithreading for all functions.
-func TestMultithreadingCycleFun(t *testing.T) {
+// TestMultithreading
+func TestMultithreading(t *testing.T) {
 	pats := generateRandomPatterns(100, 42)
 	// launch many goroutines to increase odds of finding race conditions
-	netS, netP := buildIdenticalNetworks(t, pats, 16, 16, 16)
+	netS, netP := buildIdenticalNetworks(t, pats, 2)
```
[the remark this answers is not in the document]
Comment (rcoreilly) [reply axon/threads_test.go:26]: sorry that was changed during an intermediate step and should be changed back to 16!

**Comment by rcoreilly** [comment]
There is a lot of discussion on performance issues related to this on slack thread -- I'll tag you on that so you can catch up on that.

File: axon/threads_test.go (hunk 1 above)
Comment (siboehm) [reply axon/threads_test.go:160]: @rcoreilly Can I bring these back?

Comment (rcoreilly) [reply axon/threads_test.go:160]: There is no separate parameterization for sendspike vs. anything else here -- all threading using single NThreads param.  Everything is using neuron level parallel and I didn't see any speed diffs for the diff functions, and the complexity cost is significant, so much simpler to use one NThreads.

File: axon/threads.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (siboehm) [reply axon/threads.go:22]: But back then we had much simpler Ca processing didn't we?

**Pull request merged**